### PR TITLE
Add mcxc242 base enablement

### PR DESCRIPTION
**Pull request by michal-smola**
Add basic enablement of mcxc 242, 142 and 141 socs in device tree.
Add basic enablement of mcxc series.
Add basic enablement of frdm_mcxc242 board.
Fix build error caused by uart_mcux_lpuart.c driver

### CODE REVIEW

File: boards/nxp/frdm_mcxc242/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#
+# Copyright 2024 NXP
+#
+# SPDX-License-Identifier: Apache-2.0
+#
```
Comment (nordicjm) [inline boards/nxp/frdm_mcxc242/CMakeLists.txt:5]: empty file can go

File: boards/nxp/frdm_mcxc242/doc/frdm_mcxc242.jpg
Comment (nordicjm) [inline boards/nxp/frdm_mcxc242/doc/frdm_mcxc242.jpg:1]: convert to wepb and put through https://tinypng.com/

File: soc/nxp/mcx/mcxc/Kconfig [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+config MCXC_OSC_LOW_POWER
+	bool "Low power oscillator"
+	help
+	  Set this option to use the oscillator in low-power mode.
+
+config MCXC_OSC_HIGH_GAIN
+	bool "High gain oscillator"
+	help
+	  Set this option to use the oscillator in high-gain mode.
```
Comment (nordicjm) [inline soc/nxp/mcx/mcxc/Kconfig:48]: move to dts

File: soc/nxp/mcx/mcxc/Kconfig [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+endchoice
+
+config OSC_XTAL0_FREQ
+	int "External oscillator frequency"
+	help
+	  Set the external oscillator frequency in Hz. This should be set by the
+	  board's defconfig.
```
Comment (nordicjm) [inline soc/nxp/mcx/mcxc/Kconfig:56]: dts

File: soc/nxp/mcx/mcxc/Kconfig [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+	  Selects the amount to divide down the low-frequency internal
+	  reference clock as exponent of power of 2.
+
+config MCG_LIRCDIV2
+	int "Low-frequency internal reference clock divider 2"
+	range 0 7
+	default 0
+	help
+	  Selects the amount to divide down the low-frequency internal
+	  reference clock as exponent of power of 2.
```
Comment (nordicjm) [inline soc/nxp/mcx/mcxc/Kconfig:81]: dts

File: soc/nxp/mcx/mcxc/Kconfig [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+	  devices. For program flash only devices, this byte is reserved.
+
+config MCXC_FLASH_CONFIG_FDPROT
+	hex "Data flash protection byte (FDPROT)"
+	range 0 0xff
+	default 0xff
+	help
+	  Configures the reset value of the FDPROT register for FlexNVM
+	  devices. For program flash only devices, this byte is reserved.
+
```
Comment (nordicjm) [inline soc/nxp/mcx/mcxc/Kconfig:131]: dts for anything that is hardware etc.

**Comment by dleach02** [comment]
@michal-smola I've updated your PR to remove west.yml change now that we have sdk2.16 merged

File: dts/arm/nxp/nxp_mcxc_common.dtsi
Comment (decsny) [inline dts/arm/nxp/nxp_mcxc_common.dtsi:1]: are the contents of this file actually going to be common to all MCXC down the line like the name implies?

File: include/zephyr/dt-bindings/clock/kinetis_sim.h [hunk 6]
```diff
@@ -15,6 +15,9 @@
 #define KINETIS_SIM_DMAMUX_CLK  KINETIS_SIM_BUS_CLK
 #define KINETIS_SIM_DMA_CLK     KINETIS_SIM_CORESYS_CLK
 #define KINETIS_SIM_SIM_SOPT7   7
+#define KINETIS_SIM_OSCERCLK    8
+#define KINETIS_SIM_MCGIRCLK    12
```
Comment (decsny) [inline include/zephyr/dt-bindings/clock/kinetis_sim.h:19]: these changes seem to be in the wrong commit.

File: boards/nxp/frdm_mcxc242/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (decsny) [inline boards/nxp/frdm_mcxc242/CMakeLists.txt:1]: why add an empty file with only a license?

File: drivers/serial/uart_mcux_lpuart.c [hunk 8]
```diff
@@ -1092,13 +1092,17 @@ static int mcux_lpuart_configure_init(const struct device *dev, const struct uar
 
 	LPUART_Init(config->base, &uart_config, clock_freq);
 
+#if defined(FSL_FEATURE_LPUART_HAS_MODEM_SUPPORT) && \
+	FSL_FEATURE_LPUART_HAS_MODEM_SUPPORT
```
Comment (decsny) [inline drivers/serial/uart_mcux_lpuart.c:1096]: the ci check wants this to be on one line

**Review by decsny — COMMENTED** [review]
reviewed the dts, board, and driver commits

File: dts/arm/nxp/nxp_mcxc_common.dtsi
Comment (michal-smola) [reply dts/arm/nxp/nxp_mcxc_common.dtsi:1]: Yes, the plan is to have common file to all MCXC. There is about 90% or even greater overlap of IPs for MCXC SOCs.

File: include/zephyr/dt-bindings/clock/kinetis_sim.h (hunk 6 above)
Comment (michal-smola) [reply include/zephyr/dt-bindings/clock/kinetis_sim.h:19]: It is intentional. The defines can be used to select clock source for LPUART in devicetree. Do you prefer to have this change in another commit?

File: boards/nxp/frdm_mcxc242/CMakeLists.txt (hunk 7 above)
Comment (michal-smola) [reply boards/nxp/frdm_mcxc242/CMakeLists.txt]: Removed the file .

File: drivers/serial/uart_mcux_lpuart.c (hunk 8 above)
Comment (michal-smola) [reply drivers/serial/uart_mcux_lpuart.c:1096]: CI check does not seem to fail. There has been one more instance of the same line in the file before my change already. Unified style of the file would be nice. Can I leave the added lines as is, or do you prefer to change all occurrences to the one line format?

File: boards/nxp/frdm_mcxc242/CMakeLists.txt (hunk 1 above)
Comment (michal-smola) [reply boards/nxp/frdm_mcxc242/CMakeLists.txt]: the file was removed

File: boards/nxp/frdm_mcxc242/doc/frdm_mcxc242.jpg
Comment (michal-smola) [reply boards/nxp/frdm_mcxc242/doc/frdm_mcxc242.jpg:1]: Done

**Comment by michal-smola** [comment]
I've reworked MCG, FTFA and OSC configuration from Kconfig to Devicetree to address comments by @nordicjm. Could you review it please?

File: soc/nxp/mcx/mcxc/Kconfig (hunk 2 above)
Comment (nordicjm) [reply soc/nxp/mcx/mcxc/Kconfig]: fix this in the original commit not adding something wrong then changing to to be right in a later commit

**Review by nordicjm — DISMISSED** [review]

**Comment by michal-smola** [comment]
I've reorganized the commits. Commits with dts bindings additions first, then the original commits fixed.

**Review by dleach02 — DISMISSED** [review]

**Comment by dleach02** [comment]
@nordicjm can you verify if the change addressed your request?

File: dts/bindings/clock/nxp,mcxc-osc.yaml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+description: NXP MCXC oscillator
+
+compatible: "nxp,mcxc-osc"
+
+include: fixed-clock.yaml
+
+properties:
+  load-capacitance:
+    type: int
+    enum: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
```
Comment (nordicjm) [inline dts/bindings/clock/nxp,mcxc-osc.yaml:13]: @gmarull should this be like this or just have the pf values directly?

File: dts/bindings/clock/nxp,mcxc-osc.yaml [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+description: NXP MCXC oscillator
+
+compatible: "nxp,mcxc-osc"
+
+include: fixed-clock.yaml
+
+properties:
+  load-capacitance:
```
Comment (gmarull) [inline dts/bindings/clock/nxp,mcxc-osc.yaml:11]: please, add units here

**Review by gmarull — CHANGES_REQUESTED** [review]

File: dts/bindings/clock/nxp,mcxc-osc.yaml (hunk 10 above)
Comment (michal-smola) [reply dts/bindings/clock/nxp,mcxc-osc.yaml]: The property is not load capacitance directly, but rather configuration of the load. Changed the misleading property name to load-configuration (without a unit). Meaning is explained in the description of the property. @gmarull please review.

Comment (gmarull) [reply dts/bindings/clock/nxp,mcxc-osc.yaml]: Then this is against bindings good practices, use meaningful units.

Comment (michal-smola) [reply dts/bindings/clock/nxp,mcxc-osc.yaml]: Changed to femtofarads.

Comment (michal-smola) [reply dts/bindings/clock/nxp,mcxc-osc.yaml]: @gmarull could you review the changes, please?

**Review by dleach02 — DISMISSED** [review]

**Comment by dleach02** [comment]
@gmarull can you check to see if your change request has been addressed?

File: dts/bindings/flash_controller/nxp,kinetis-ftfa.yaml [hunk 11]
```diff
@@ -3,3 +3,22 @@ description: NXP Kinetis Flash Memory Module A (FTFA)
...
+  fopt:
+    type: int
+    description: |
+      Configures the reset value of the FOPT register, which includes boot,
+      NMI, and EzPort options.
+
+  config-field-offset:
+    type: int
+    description: |
+      "Flash configuration field offset"
```
Comment (gmarull) [inline dts/bindings/flash_controller/nxp,kinetis-ftfa.yaml:24]: redundant `"`

File: dts/bindings/clock/nxp,mcxc-osc.yaml [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+      - 12000
+      - 14000
+      - 16000
+      - 18000
+      - 20000
+      - 22000
+      - 24000
+      - 26000
+      - 28000
+      - 30000
```
Comment (gmarull) [inline dts/bindings/clock/nxp,mcxc-osc.yaml:29]: why are you using femto units when all values seem to be scaled by 1000?

**Review by gmarull — DISMISSED** [review]

File: dts/bindings/clock/nxp,mcxc-osc.yaml (hunk 12 above)
Comment (michal-smola) [reply dts/bindings/clock/nxp,mcxc-osc.yaml:29]: There is no apparent reason. I will change it to picofarads.

Comment (michal-smola) [reply dts/bindings/clock/nxp,mcxc-osc.yaml]: Changed to picofarads. @gmarull could you review, please?

File: dts/bindings/flash_controller/nxp,kinetis-ftfa.yaml (hunk 11 above)
Comment (michal-smola) [reply dts/bindings/flash_controller/nxp,kinetis-ftfa.yaml]: Removed redundant "

**Review by dleach02 — DISMISSED** [review]

File: soc/nxp/mcx/mcxc/pinctrl_soc.h
Comment (danieldegrasse) [inline soc/nxp/mcx/mcxc/pinctrl_soc.h:1]: This file mirrors the kinetis common pin control file- is there a reason we can't include definitions from there?
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h

File: drivers/serial/uart_mcux_lpuart.c (hunk 8 above)
Comment (danieldegrasse) [reply drivers/serial/uart_mcux_lpuart.c:1096]: This commit should come before the commits adding support for the MCXC242, if that is the SOC causing this build error

File: dts/bindings/flash_controller/nxp,kinetis-ftfa.yaml
Comment (danieldegrasse) [inline dts/bindings/flash_controller/nxp,kinetis-ftfa.yaml:1]: The commit message here says:

> Some FTFA properties are configured in Kconfig. Add the properties in
dts binging as the properties configure harware.


Is this correct? I don't see where these properties are being moved from Kconfig to DTS. Perhaps the commit message needs to be something like the following?

> Add configuration settings for the FOPT, FSEC, and flash configuration offset to the FTFA module binding

File: include/zephyr/dt-bindings/clock/kinetis_sim.h (hunk 6 above)
Comment (danieldegrasse) [reply include/zephyr/dt-bindings/clock/kinetis_sim.h:19]: Adding here- please do move this change to another commit

File: dts/bindings/flash_controller/nxp,kinetis-ftfa.yaml
Comment (michal-smola) [reply dts/bindings/flash_controller/nxp,kinetis-ftfa.yaml:1]: Commit message was changed as proposed.

File: include/zephyr/dt-bindings/clock/kinetis_sim.h (hunk 6 above)
Comment (michal-smola) [reply include/zephyr/dt-bindings/clock/kinetis_sim.h:19]: Moved to separate commit.

File: drivers/serial/uart_mcux_lpuart.c (hunk 8 above)
Comment (michal-smola) [reply drivers/serial/uart_mcux_lpuart.c:1096]: Moved the commit.

File: dts/bindings/clock/nxp,kinetis-mcg.yaml [hunk 13]
```diff
@@ -14,5 +14,19 @@ properties:
...
+      Internal Reference Clock Divider.
+      Division factor is given as 2^fcrdiv.
+      The valid range is 0 to 7.
+
+  lircdiv2:
+    type: int
+    description: |
+      Second Low-frequency Internal Reference Clock Divider.
+      Division factor is given as 2^lircdiv2.
+      The valid range is 0 to 7.
```
Comment (nordicjm) [inline dts/bindings/clock/nxp,kinetis-mcg.yaml:29]: why can't these be enums?

File: dts/bindings/flash_controller/nxp,kinetis-ftfa.yaml [hunk 14]
```diff
@@ -3,3 +3,22 @@ description: NXP Kinetis Flash Memory Module A (FTFA)
...
 
 include: flash-controller.yaml
+
+properties:
+  fsec:
+    type: int
+    description: |
+      Configures the reset value of the FSEC register, which includes
+      backdoor key access, mass erase, factory access, and flash security
+      options.
```
Comment (nordicjm) [inline dts/bindings/flash_controller/nxp,kinetis-ftfa.yaml:13]: seems like these could be enums too?

File: boards/nxp/frdm_mcxc242/doc/index.rst [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+- `MCX C24x Datasheet`_
+- `MCX C24x Reference Manual`_
+- `FRDM-MCXC242 Website`_
+- `FRDM-MCXC242 User Manual`_
+- `FRDM-MCXC242 Design Files`_
+
+Supported Features
+==================
+
+The FRDM-MCXC242 board configuration supports the following hardware features:
```
Comment (nordicjm) [inline boards/nxp/frdm_mcxc242/doc/index.rst:49]: use west board name with \`\` around it and *board target

File: boards/nxp/frdm_mcxc242/frdm_mcxc242_defconfig [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+# Copyright 2024 NXP
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+CONFIG_SERIAL=y
+CONFIG_GPIO=y
+CONFIG_PINCTRL=y
```
Comment (nordicjm) [inline boards/nxp/frdm_mcxc242/frdm_mcxc242_defconfig:11]: no pinctrl selection at board level

**Review by nordicjm — CHANGES_REQUESTED** [review]
last 2 comments are blockers, first 2 are optional

File: dts/bindings/clock/nxp,kinetis-mcg.yaml (hunk 13 above)
Comment (michal-smola) [reply dts/bindings/clock/nxp,kinetis-mcg.yaml]: @nordicjm, thank you for review. Changed to enum.

File: dts/bindings/flash_controller/nxp,kinetis-ftfa.yaml (hunk 14 above)
Comment (michal-smola) [reply dts/bindings/flash_controller/nxp,kinetis-ftfa.yaml:13]: Prefered to be int.

File: boards/nxp/frdm_mcxc242/doc/index.rst (hunk 15 above)
Comment (michal-smola) [reply boards/nxp/frdm_mcxc242/doc/index.rst]: Fixed.

File: boards/nxp/frdm_mcxc242/frdm_mcxc242_defconfig (hunk 16 above)
Comment (michal-smola) [reply boards/nxp/frdm_mcxc242/frdm_mcxc242_defconfig]: Removed pinctrl selection

**Review by nordicjm — APPROVED** [review]

**Review by danieldegrasse — APPROVED** [review]

**Review by mmahadevan108 — APPROVED** [review]

**Pull request merged**